src/sessions.ts

```diff
@@ -591,18 +589,18 @@ function attemptTransaction<TSchema>(
 
   if (!isPromiseLike(promise)) {
     session.abortTransaction().catch(() => null);
-    throw new MongoInvalidArgumentError(
-      'Function provided to `withTransaction` must return a Promise'
+    return Promise.reject(
+      new MongoInvalidArgumentError('Function provided to `withTransaction` must return a Promise')
     );
   }
 
   return promise.then(
-    () => {
+    result => {
       if (userExplicitlyEndedTransaction(session)) {
-        return;
+        return result;
       }
 
-      return attemptTransactionCommit(session, startTime, fn, options);
+      return attemptTransactionCommit(session, startTime, fn, options).then(() => result);
```

Comment:
What happens if we need to retry the transaction?  I think we'd get incorrect behavior, because the first call to `attemptTransaction` would always return the result of the first call to `fn` - but I'd assume that we want to return the `result` associated with the successful `attemptTransaction`.

Reply:
Perhaps this is okay and expected behavior?  We document that `Client.withSession` may call its callback more than once - we should here too at the very least.  But I think since we may call the callback twice, it's an acceptable limitation that the callback should be idempotent.

Reply:
edit: it is noted for this method

Reply:
pending slack convo

Reply:
Updated to return the last result and added a test